statsmodels/doe/latin.py

```diff
@@ -0,0 +1,200 @@
+    toward 2D and 3D subprojections. Distance based design better space filling
+    but less robust to subprojections.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension of the parameter space.
+    n_sample : int
+        Number of samples to generate in the parametr space.
+    bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+        Desired range of transformed data. The transformation apply the bounds
+        on the sample and not the theoretical space, unit cube. Thus min and
+        max values of the sample will coincide with the bounds.
+    start_design : array_like (n_samples, k_vars)
+        Initial design of experiment to optimize.
+
+    Returns
+    -------
+    sample : array_like (n_samples, k_vars)
+        Optimal Latin hypercube Sampling.
+
+    References
+    ----------
+    [1] Damblin et al., "Numerical studies of space filling designs:
+    optimization of Latin Hypercube Samples and subprojection properties",
+    Journal of Simulation, 2013.
+
+    """
+    doe = start_design
+    if doe is None:
+        doe = orthogonal_latin_hypercube(dim, n_sample, bounds)
+
+    def _perturb_doe(x, sample, bounds):
+        """Disturbe the Design of Experiment.
+
+        Parameters
+        ----------
+        x : list of int
+            It is a list of:
+                idx : int
+                    Index value of the components to compute
+        sample : array_like (n_samples, k_vars)
+            Sample to perturb.
+        bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+            Desired range of transformed data. The transformation apply the
+            bounds on the sample and not the theoretical space, unit cube. Thus
+            min and max values of the sample will coincide with the bounds.
+
+        Returns
+        -------
+        discrepancy : float
+            Centered discrepancy.
+
+        """
+        col, row_1, row_2 = map(int, x)
+        doe[row_1, col], doe[row_2, col] = doe[row_2, col], doe[row_1, col]
+
+        return discrepancy(sample, bounds)
+
+    # Total number of possible design
+    complexity = dim * n_sample ** 2
+
+    if have_basinhopping and ((complexity > 1e6) or force):
+        bounds_optim = ([0, dim - 1], [0, n_sample - 1], [0, n_sample - 1])
+    else:
+        bounds_optim = (slice(0, dim - 1, 1), slice(0, n_sample - 1, 1),
+                        slice(0, n_sample - 1, 1))
+
+    for n in range(niter):
+        if have_basinhopping and ((complexity > 1e6) or force):
+            minimizer_kwargs = {"method": "L-BFGS-B", "bounds": bounds_optim,
+                                "args": (doe, bounds)}
+            optimum = basinhopping(_perturb_doe, [0, 0, 0], niter=100,
+                                   minimizer_kwargs=minimizer_kwargs).x
+        else:
+            optimum = brute(_perturb_doe, ranges=bounds_optim,
+                            finish=fmin, args=(doe, bounds))
```

Comment:
Oh didn’t know, even with using `l-bfgs-b`?

Reply:
l-bfgs-b will use the bounds. I'm not sure which optimizers in scipy only try to evaluate points within bounds during the search (*) but they are supposed to return values within the bounds.

(*) there have been modifications for this in scipy across version, but I didn't pay close attention because so far we almost never use constrained solvers in statsmodels.

I doubt l-bfgs-b will work well because it is looking for continuous parameters and locally the perturb_doe function is flat because of the integer truncation, so gradient and hessian approximation in l-bfgs-b or similar will not help.

There is no good optimizer for discrete parameter spaces in scipy, AFAIK (unless there is enough continuity or smoothness in the extension to real numbers.)

Reply:
I agree with the reasoning so I will set `finish=None`